cmd/snap-update-ns/utils.go

```diff
@@ -144,6 +144,46 @@ func secureMkDir(fd int, segments []string, i int, perm os.FileMode, uid, gid in
 }
 
 // secureMkdirAll is the secure variant of os.MkdirAll.
+func secureMkFile(fd int, segments []string, i int, perm os.FileMode, uid, gid int) error {
+	logger.Debugf("secure-mk-file %d %q %d %v %d %d", fd, segments, i, perm, uid, gid)
+	segment := segments[i]
+	made := true
+	var newFd int
+	var err error
+
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_RDWR
+
+	// Open the final path segment as a file. Try to create the file (so that
+	// we know if we need to chown it) but fall back to just opening an
+	// existing one.
+
+	newFd, err = sysOpenat(fd, segment, openFlags|syscall.O_CREAT|syscall.O_EXCL, uint32(perm.Perm()))
+	if err != nil {
+		switch err {
+		case syscall.EEXIST:
+			// If the file exists then just open it without O_EXCL
+			newFd, err = sysOpenat(fd, segment, openFlags, 0)
+			if err != nil {
+				return fmt.Errorf("cannot open file %q: %v", segment, err)
+			}
+			made = false
+		default:
+			return fmt.Errorf("cannot open file %q: %v", segment, err)
+		}
+	}
+	defer sysClose(newFd)
+
+	if made {
+		// Chown the file if we made it.
+		if err := sysFchown(newFd, uid, gid); err != nil {
+			return fmt.Errorf("cannot chown file %q to %d.%d: %v", segment, uid, gid, err)
+		}
+	}
+
+	return nil
```

Comment:
Why aren't you returning newFd here like with the secureMkdir? This is an API inconsistency that at the very least should be documented up above.

Reply:
Because the point of the API (at least for what I need it for) is to ensure the directory / file exists and there is no expected follow up. We don't write to those files, just use them as mount points.